Data/HashMap/Base.hs

```diff
@@ -181,17 +184,17 @@ data HashMap k v
     | BitmapIndexed !Bitmap !(A.Array (HashMap k v))
     | Leaf !Hash !(Leaf k v)
     | Full !(A.Array (HashMap k v))
-    | Collision !Hash !(A.Array (Leaf k v))
+    | Collision !Hash !(Leaf k v) !(Leaf k v) !(HashMap k v)
       deriving (Typeable)
```

Comment:
I think you've missed two invariants (because they're not documented with the type):

1. An array never contains an empty map.
2. An array is never empty.
3. Only a transient array can have exactly one element. These transient arrays are only used to reduce the amount of extra code on collisions.

Reply:
You're talking about the `A.Array` that I just deleted, right? So these invariants are not necessary anymore?.
Or are you talking about the `BitmapIndexed` constructor?

Reply:
Huh? No... I'm saying you can simplify the code for both `BitmapIndexed` and `Full` by using those invariants, but also that you need to maintain them. Also, if the test suite doesn't check for them, it should--unless I'm wrong and we don't really maintain them.